### PR TITLE
BUGZ-666: Toolbar Artifact Always Present

### PR DESCRIPTION
**Pull request by PrestonB1123**
A More Permanent fix for this bug.
https://highfidelity.atlassian.net/browse/BUGZ-666

### CODE REVIEW

File: libraries/ui/src/OffscreenUi.h [hunk 1]
```diff
@@ -270,7 +271,8 @@ private slots:
     QList<QObject*> _modalDialogListeners;
     std::unordered_map<int, bool> _pressedKeys;
     VrMenu* _vrMenu { nullptr };
-    QQueue<std::function<void(VrMenu*)>> _queuedMenuInitializers; 
+    QQueue<std::function<void(VrMenu*)>> _queuedMenuInitializers;
+    ToolbarProxy* _currentToolbarProxy;
```
Comment (zfox23) [inline libraries/ui/src/OffscreenUi.h]: please remove this change - we don't need the `_currentToolbarProxy` member anymore!

File: libraries/ui/src/OffscreenUi.h [hunk 2]
```diff
@@ -24,6 +24,7 @@
 #include <DependencyManager.h>
 
 #include "ui/OffscreenQmlSurface.h"
+#include "ui/ToolbarScriptingInterface.h"
```
Comment (zfox23) [inline libraries/ui/src/OffscreenUi.h]: please move this include to the cpp file

File: libraries/ui/src/ui/ToolbarScriptingInterface.cpp [hunk 3]
```diff
@@ -119,4 +119,4 @@ ToolbarProxy* ToolbarScriptingInterface::getToolbar(const QString& toolbarId) {
     }
 
     return new ToolbarProxy(rawToolbar);
-}
+}
```
Comment (zfox23) [inline libraries/ui/src/ui/ToolbarScriptingInterface.cpp]: please revert this change :)

File: scripts/simplifiedUI/ui/simplifiedUI.js [hunk 4]
```diff
@@ -457,16 +457,16 @@ function onGeometryChanged(rect) {
...
-        Script.setTimeout(function () {
-            var toolbar = Toolbars.getToolbar(TOOLBAR_NAME);
-            toolbar.writeProperty("visible", false);
-        }, TIMEOUT_BEFORE_REHIDE_TOOLBAR_MS);
+    }
+}
+
+function onToolbarVisibleChanged(isVisible, toolbarName) {
+    var toolbar = Toolbars.getToolbar(toolbarName);
+    if (isVisible) {
```
Comment (zfox23) [inline scripts/simplifiedUI/ui/simplifiedUI.js]: you also need to make sure that `toolbarName` is the one you care about (defined as `TOOLBAR_NAME` in this file IIRC)

File: scripts/simplifiedUI/ui/simplifiedUI.js [hunk 5]
```diff
@@ -457,16 +457,16 @@ function onGeometryChanged(rect) {
...
-        // works for now, but not a permanent fix by any means.
-        Script.setTimeout(function () {
-            var toolbar = Toolbars.getToolbar(TOOLBAR_NAME);
-            toolbar.writeProperty("visible", false);
-        }, TIMEOUT_BEFORE_REHIDE_TOOLBAR_MS);
+    }
+}
+
+function onToolbarVisibleChanged(isVisible, toolbarName) {
+    var toolbar = Toolbars.getToolbar(toolbarName);
```
Comment (zfox23) [inline scripts/simplifiedUI/ui/simplifiedUI.js]: you can move this into the conditional below

File: scripts/simplifiedUI/ui/simplifiedUI.js [hunk 6]
```diff
@@ -542,9 +543,7 @@ function shutdown() {
 
         if (!HMD.active) {
             var toolbar = Toolbars.getToolbar(TOOLBAR_NAME);
-            if (toolbar) {
-                toolbar.writeProperty("visible", true);
-            } 
+            toolbar.writeProperty("visible", true);
```
Comment (zfox23) [inline scripts/simplifiedUI/ui/simplifiedUI.js]: whoops! looks like you missed this when doing your merge conflict resolution. you're going to want to keep the code that was here before.

File: libraries/ui/src/OffscreenUi.h (hunk 1 above)
Comment (PrestonB1123) [reply libraries/ui/src/OffscreenUi.h]: Ack

File: libraries/ui/src/OffscreenUi.h (hunk 2 above)
Comment (PrestonB1123) [reply libraries/ui/src/OffscreenUi.h]: Ack

File: scripts/simplifiedUI/ui/simplifiedUI.js (hunk 6 above)
Comment (PrestonB1123) [reply scripts/simplifiedUI/ui/simplifiedUI.js]: Ack

File: scripts/simplifiedUI/ui/simplifiedUI.js (hunk 4 above)
Comment (PrestonB1123) [reply scripts/simplifiedUI/ui/simplifiedUI.js]: Ack

File: scripts/simplifiedUI/ui/simplifiedUI.js (hunk 5 above)
Comment (PrestonB1123) [reply scripts/simplifiedUI/ui/simplifiedUI.js]: Ack

File: libraries/ui/src/ui/ToolbarScriptingInterface.cpp (hunk 3 above)
Comment (PrestonB1123) [reply libraries/ui/src/ui/ToolbarScriptingInterface.cpp]: Ack

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19945/HighFidelity-Beta-PR15752-8fa80f8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19945/HighFidelity-Beta-Interface-PR15752-8fa80f8.dmg

File: libraries/ui/src/OffscreenUi.cpp [hunk 7]
```diff
@@ -688,6 +689,11 @@ void OffscreenUi::createDesktop(const QUrl& url) {
             menuInitializer(_vrMenu);
         }
 
+
+        QString toolbarName = newObject->property("toolbarObjectName").toString(); // will return "com.highfidelity.interface.toolbar.system"
```
Comment (zfox23) [inline libraries/ui/src/OffscreenUi.cpp]: oh wait! we don't need this either, right?

File: libraries/ui/src/OffscreenUi.h [hunk 8]
```diff
@@ -270,7 +270,7 @@ private slots:
     QList<QObject*> _modalDialogListeners;
     std::unordered_map<int, bool> _pressedKeys;
     VrMenu* _vrMenu { nullptr };
-    QQueue<std::function<void(VrMenu*)>> _queuedMenuInitializers; 
+    QQueue<std::function<void(VrMenu*)>> _queuedMenuInitializers;
```
Comment (zfox23) [inline libraries/ui/src/OffscreenUi.h]: please revert this change so the file isn't marked changed

File: scripts/simplifiedUI/ui/simplifiedUI.js [hunk 9]
```diff
@@ -457,16 +457,16 @@ function onGeometryChanged(rect) {
...
-            var toolbar = Toolbars.getToolbar(TOOLBAR_NAME);
-            toolbar.writeProperty("visible", false);
-        }, TIMEOUT_BEFORE_REHIDE_TOOLBAR_MS);
+    }
+}
+
+function onToolbarVisibleChanged(isVisible, toolbarName) {
+    if (isVisible && toolbarName == TOOLBAR_NAME) {
+        var toolbar = Toolbars.getToolbar(toolbarName);
+        toolbar.writeProperty("visible", false);
```
Comment (zfox23) [inline scripts/simplifiedUI/ui/simplifiedUI.js]: wrap this in `if (toolbar) {` please!

File: scripts/simplifiedUI/ui/simplifiedUI.js (hunk 6 above)
Comment (zfox23) [reply scripts/simplifiedUI/ui/simplifiedUI.js]: looks like you missed this one still :)

File: scripts/simplifiedUI/ui/simplifiedUI.js (hunk 9 above)
Comment (PrestonB1123) [reply scripts/simplifiedUI/ui/simplifiedUI.js]: Ack

File: libraries/ui/src/OffscreenUi.h (hunk 8 above)
Comment (PrestonB1123) [reply libraries/ui/src/OffscreenUi.h]: Ack

File: libraries/ui/src/OffscreenUi.cpp (hunk 7 above)
Comment (PrestonB1123) [reply libraries/ui/src/OffscreenUi.cpp]: Ack

File: scripts/simplifiedUI/ui/simplifiedUI.js [hunk 10]
```diff
@@ -457,16 +457,18 @@ function onGeometryChanged(rect) {
...
         Camera.setModeString("first person");
-    } else if (Settings.getValue("simplifiedUI/keepExistingUIAndScripts", false)) {
-        // works for now, but not a permanent fix by any means.
-        Script.setTimeout(function () {
-            var toolbar = Toolbars.getToolbar(TOOLBAR_NAME);
+    }
+}
+
+function onToolbarVisibleChanged(isVisible, toolbarName) {
+    if (isVisible && toolbarName == TOOLBAR_NAME) {
```
Comment (zfox23) [inline scripts/simplifiedUI/ui/simplifiedUI.js]:
```suggestion
    if (isVisible && toolbarName == TOOLBAR_NAME && Settings.getValue("simplifiedUI/keepExistingUIAndScripts", false)) {
```

Comment (PrestonB1123) [reply scripts/simplifiedUI/ui/simplifiedUI.js]: Ack

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19952/HighFidelity-Beta-PR15752-e32d6e9.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19952/HighFidelity-Quest-Beta-PR15752-e32d6e9.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19952/HighFidelity-Beta-PR15752-e32d6e9.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19952/HighFidelity-Beta-Interface-PR15752-e32d6e9.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19952/HighFidelity-Beta-PR15752-e32d6e9.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19952/HighFidelity-Beta-Interface-PR15752-e32d6e9.dmg
